### PR TITLE
Messaging Between contexts (ps and us)

### PR DESCRIPTION
**Pull request by TheColaber**
TODO:
- Test a lot
- Move things from transition.js to another file
- anything else... im pretty sure it works...

Semi-resolving #1671

### CODE REVIEW

File: addon-api/common/Self.js [hunk 1]
```diff
@@ -13,7 +13,56 @@ export default class Self extends Listenable {
...
+      name: "message",
+      addonId,
+      data: { msg, scope },
+    };
+    if (this._template) {
+      // CS
+      this._template.setAttribute(`data-fire-event__${Date.now()}`, JSON.stringify(sending));
+    } else {
+      // PS
+      chrome.tabs.query({}, (tabs) =>
```
Comment (RedGuy12) [inline addon-api/common/Self.js:42]: But... this wouldn't be available in firefox... would it?

Comment (mxmou) [reply addon-api/common/Self.js:42]: Firefox supports most `chrome.*` APIs.

Comment (TheColaber) [reply addon-api/common/Self.js:42]: We use this same function in other places as well.

File: addon-api/common/Self.js [hunk 2]
```diff
@@ -13,7 +13,56 @@ export default class Self extends Listenable {
...
+        tabs.forEach(
+          (tab) =>
+            (tab.url || (!tab.url && typeof browser !== "undefined")) &&
+            chrome.tabs.sendMessage(tab.id, {
+              fireEvent: {
+                target: "self",
+                name: "message",
+                addonId,
+                data: { msg, scope },
```
Comment (RedGuy12) [inline addon-api/common/Self.js:51]: use `sending` from L31?

File: content-scripts/inject/module.js [hunk 3]
```diff
@@ -97,11 +97,11 @@ const observer = new MutationObserver((mutationsList) => {
         const eventTarget = scratchAddons.eventTargets[attrVal.target].find(
           (eventTarget) => eventTarget._addonId === attrVal.addonId
         );
-        if (eventTarget) eventTarget.dispatchEvent(new CustomEvent(attrVal.name));
+        if (eventTarget) eventTarget.dispatchEvent(new CustomEvent(attrVal.name, { detail: attrVal.data || {} }));
       } else {
         // Global events, like auth change
         scratchAddons.eventTargets[attrVal.target].forEach((eventTarget) =>
-          eventTarget.dispatchEvent(new CustomEvent(attrVal.name))
+          eventTarget.dispatchEvent(new CustomEvent(attrVal.name, { detail: attrVal.data || {} }))
```
Comment (RedGuy12) [inline content-scripts/inject/module.js:104]: wouldn't it always be a local event?

**Comment by TheColaber** [comment]
Should be redone for new system... new pr soon:tm:

**Pull request closed without merging**